[RISCV] Add processor definition for XiangShan-KunMingHu-V2R2

liliumShade opened the pull request:
XiangShan-KunMingHu is the third generation of Open-source high-performance RISC-V processor developed by Beijing Institute of Open Source Chip (BOSC) , and its latest version is V2R2.

The KunMingHu manual is now available at https://github.com/OpenXiangShan/XiangShan-User-Guide/releases.
It will be updated on the official XiangShan documentation site: https://docs.xiangshan.cc/zh-cn/latest

You can find the corresponding ISA extension from the XiangShan Github repository: https://github.com/OpenXiangShan/XiangShan/blob/master/src/main/scala/xiangshan/Parameters.scala

If you want to track the latest performance data of KunMingHu, please check XiangShan Biweekly: https://docs.xiangshan.cc/zh-cn/latest/blog

This PR adds the processor definition for KunMingHu V2R2, developed by the XSCC team https://github.com/orgs/OpenXiangShan/teams/xscc.

The scheduling model for XiangShan-KunMingHu V2R2 will be submitted in a subsequent PR.

liliumShade committed:
[RISCV] Add processor definition for XiangShan-KunMingHu-V2R2

Co-Authored-By: Shenglin Tang <tangshenglin@ict.ac.cn>
Co-Authored-By: Xu, Zefan <ceba_robot@outlook.com>
Co-Authored-By: Tang Haojin <tanghaojin@outlook.com>

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVProcessors.td:
```diff
@@ -553,6 +553,37 @@ def XIANGSHAN_NANHU : RISCVProcessorModel<"xiangshan-nanhu",
+def XIANGSHAN_KUNMINGHU : RISCVProcessorModel<"xiangshan-kunminghu",
+                                          NoSchedModel,
+                                          !listconcat(!listremove(RVA23S64Features,
+                                                      [FeatureStdExtZiccamoa,
```
Any plan to make KMH RVA23-compatible?

tangshenglin replied:
V2R2 is a validated and stable release. Full support for RVA23 is also in the pipeline, with updates expected to be completed by 2025. The timeline will be dynamically adjusted based on progress. Once finalized, the XSCC team will synchronize updates accordingly.

wangpc-pp replied:
Then I think we should wait for the freeze of features. I don't think we will accept an intermediate version.

Tang-Haojin replied:
> Then I think we should wait for the freeze of features. I don't think we will accept an intermediate version.

Hi @wangpc-pp, the KunMingHu-V2R2 version of the ISA strings has been frozen, and support for the "ziccamoa", "zihintntl", and "zawrs" extensions has also been updated. Please check https://github.com/OpenXiangShan/XiangShan/pull/4219 (This PR has only updated the isa string, and the support of these three extensions had been done before this PR).

liliumShade committed:
Add "ziccamoa"  "zihintntl" and "zawrs"

Xiangshan-Kunminghu now compatible with RVA23S64 specification

Co-Authored-By: Shenglin Tang <tangshenglin@ict.ac.cn>
Co-Authored-By: Xu, Zefan <ceba_robot@outlook.com>
Co-Authored-By: Tang Haojin <tanghaojin@outlook.com>

liliumShade committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-V2R2

liliumShade committed:
Add XIANGSHAN-KUNMINGHU CPU test

Verify 24 target features for `-mcpu=xiangshan-kunminghu` including:
- Standard extensions (h/zfa/zvbb/sstc/sscofpmf)
- Vendor-specific extensions (shgatpa/shvsatpa/ssu64xl)

liliumShade committed:
Merge branch 'OpenXiangShan/KunMingHu-V2R2' of https://github.com/OpenXiangShan/llvm-project into OpenXiangShan/KunMingHu-V2R2

liliumShade committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-V2R2

liliumShade commented:
> I think some feature is missing in `clang/test/Driver/riscv-cpus.c`, such as "Zvbb".

 Thanks for the review! We've added the Zvbb test case and extended validation for XIANGSHAN-KUNMINGHU CPU's target features (including privileged/custom extensions)

tclin914 commented on llvm/lib/Target/RISCV/RISCVProcessors.td:
```diff
@@ -558,6 +558,34 @@ def XIANGSHAN_NANHU : RISCVProcessorModel<"xiangshan-nanhu",
                                             TuneShiftedZExtWFusion]>;
 
+def XIANGSHAN_KUNMINGHU : RISCVProcessorModel<"xiangshan-kunminghu",
+                                          NoSchedModel,
```
indentation alignment

liliumShade replied:
Done!

liliumShade committed:
FIx indentation in RISCVProcessors.td

This corrects whitespace alignment in the xiangshan-kunminghu part of RISCVProcessors.td.

liliumShade committed:
Merge branch 'OpenXiangShan/KunMingHu-V2R2' of https://github.com/OpenXiangShan/llvm-project into OpenXiangShan/KunMingHu-V2R2

wangpc-pp commented on llvm/docs/ReleaseNotes.md:
```diff
@@ -126,6 +126,7 @@ Changes to the PowerPC Backend
 Changes to the RISC-V Backend
 -----------------------------
 
+
```
Remove this extra blank line.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVProcessors.td:
```diff
@@ -558,6 +558,34 @@ def XIANGSHAN_NANHU : RISCVProcessorModel<"xiangshan-nanhu",
+def XIANGSHAN_KUNMINGHU : RISCVProcessorModel<"xiangshan-kunminghu",
+                                              NoSchedModel,
+                                              !listconcat(RVA23S64Features,
+                                              [FeatureStdExtZicsr,
```
Zicsr is in `RVA23S64Features` already.

liliumShade replied:
You are right, but I didn't find FeatureStdExtZicsr in RISCVProfiles.td. So, just to confirm, there won't be any problem if it is removed, right? 😮

wangpc-pp replied:
Yes you can remove it because the F/D extensions imply `Zicsr`. I can add it explicitly to the RISCVProfiles.td to avoid future misleadings.

wangpc-pp replied:
https://github.com/llvm/llvm-project/pull/136134

liliumShade committed:
Remove blank line in ReleaseNotes.md

liliumShade committed:
Remove "FeatureStdExtZicsr" from RISCVProcessors.td

wangpc-pp commented:
LGTM in general, but I have a question here: can you clarify the naming strategy? The name used in `-mcpu` is `xiangshan-kunminghu`, which corresponds to the `V2R2` version now apparently. Then, will there be `V2R3`/`V3R2`/...? If so, what should we use in `-mcpu`?

liliumShade commented:
> LGTM in general, but I have a question here: can you clarify the naming strategy? The name used in `-mcpu` is `xiangshan-kunminghu`, which corresponds to the `V2R2` version now apparently. Then, will there be `V2R3`/`V3R2`/...? If so, what should we use in `-mcpu`?
  
Thank you for raising this! To clarify: V2R2 is the current frozen version of kunminghu, the subsequent development in the plan will not affect the ISA string and pipeline.

If significant changes ever arise in a future major version (e.g., a hypothetical V3), we will introduce a new -mcpu name to reflect that divergence.

wangpc-pp reviewed (approved):
LGTM! Thanks for the insistence!

liliumShade committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-V2R2
